### DISCUSSION TITLE
typegen fails silently

### DISCUSSION THREAD

**Discussion by EightArmCode** [statelyai/xstate#3391] [Q&A] [answered]
Hello! Thank you for xstate. So far we're very happy with it.

I recently ran into a problem where typegen doesn't generate an output file on my latest branch. I'm assuming this is because I'm changing from passing an object directly to `createMachine` to building the object and then passing it in later, and I've missed something in the config or typing. When I use the CLI `--help flag`, I find that there's only the generator and the watcher commands, but I was hoping there might be a trick to debugging that I haven't come across yet.

I can't really share code, because it's proprietary. For now just looking for a debug option. If there's an error running the script, should there be console output for it?

TIA!

**Comment by Andarist** [comment, marked as answer]
It's true that we won't be able to generate typegen data for cases like this:
```js
import { createMachine } from 'xstate'

const config = {}
createMachine(config)
```

We should probably think about ESlint rules that would catch this and maybe we could also implement a simple linter into the CLI itself. You could create a feature request for those here: https://github.com/statelyai/xstate-tools

**Reply by EightArmCode** [reply]
Thank you for the quick response. That's helpful to know that won't work. I just came across the `createModel` function, and I'm hoping this will improve our visibility and type safety. My initial question however, was about an error output from the typegen. I'm assuming that's not available yet either? Are there any plans in the works for this?

**Reply by EightArmCode** [reply]
I'm also wondering - will the typegen work if we use `createModel` and then create the machine with `myModel.createMachine`?

**Reply by EightArmCode** [reply]
I'm trying it out, and it seems to be working.

**Reply by Andarist** [reply]
> I just came across the createModel function, and I'm hoping this will improve our visibility and type safety. 

We don't recommend using `createModel` - typegen supersedes it.

> My initial question however, was about an error output from the typegen. I'm assuming that's not available yet either?

Could you give a concrete example of errors that could be surfaced and which are not visible anywhere today?

**Reply by EightArmCode** [reply]
That's a good question. Perhaps this is not perceived as an error by the stately team, but it took me a while to figure out that typegen was not creating an output file in our CI, and then it was not immediately apparent as to why.

Incidentally, will we be able to pass in a config object to createMachine, and have that play nicely with typegen?

>We don't recommend using createModel - typegen supersedes it.

Is there a working example of the most current preferred method for instantiating a machine with type safety?

**Reply by Andarist** [reply]
> Perhaps this is not perceived as an error by the stately team, but it took me a while to figure out that typegen was not creating an output file in our CI, and then it was not immediately apparent as to why.

Yeah, it's not an error per-se - if we are talking about the situation described above. However, we should find ways to surface those kinds of problems to devs (that's why I've suggested creating an ESLint plugin for this).

> Incidentally, will we be able to pass in a config object to createMachine, and have that play nicely with typegen?

I'm unsure what do you mean - could you rephrase?

> Is there a working example of the most current preferred method for instantiating a machine with type safety?

Basically, this combined with VSCode extension or the CLI:
```ts
import { createMachine } from 'xstate'

// argument has to be an object literal
createMachine({
  tsTypes: {}, // this makes it clear for our tooling that you want typegen metadata to be generated
  schema: {
    context: {} as MyContext,
    events: {} as MyEvent,
  }
})
```

**Reply by EightArmCode** [reply]
"Incidentally, will we be able to pass in a config object to createMachine, and have that play nicely with typegen?"

>I'm unsure what do you mean - could you rephrase?

The scenario you said that doesn't work with typegen, i.e. 

```js
import { createMachine } from 'xstate'

const config = {}
createMachine(config)
```

I'm wondering if we can expect this to work with typegen in the future? For now I'm thinking I'll do something like this:

```js
import { createMachine } from 'xstate'
import { events } from './myEventsFile'

createMachine({
   ...baseMachine,
   events,
})
```
